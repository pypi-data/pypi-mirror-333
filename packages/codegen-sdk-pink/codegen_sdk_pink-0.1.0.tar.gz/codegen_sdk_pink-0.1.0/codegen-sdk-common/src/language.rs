use std::{hash::Hash, num::NonZeroU16, path::PathBuf, sync::Arc};

use convert_case::{Case, Casing};
use mockall::automock;
use proc_macro2::Span;
use tree_sitter::Parser;

use crate::{
    errors::ParseError,
    naming::normalize_type_name,
    parser::{Node, parse_node_types},
};
#[derive(Debug, Eq, PartialEq)]
pub struct Language {
    name: &'static str,
    pub struct_name: &'static str,
    pub node_types: &'static str,
    pub file_extensions: &'static [&'static str],
    tree_sitter_language: tree_sitter::Language,
    pub tag_query: &'static str,
    nodes: Vec<Arc<Node>>,
}
impl Hash for Language {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.name.hash(state);
    }
}

#[automock]
impl Language {
    pub fn new(
        name: &'static str,
        struct_name: &'static str,
        node_types: &'static str,
        file_extensions: &'static [&'static str],
        tree_sitter_language: tree_sitter::Language,
        tag_query: &'static str,
    ) -> anyhow::Result<Self> {
        let nodes = parse_node_types(node_types)?
            .into_iter()
            .map(|node| Arc::new(node))
            .collect();
        Ok(Self {
            name,
            struct_name,
            node_types,
            file_extensions,
            tree_sitter_language,
            tag_query,
            nodes,
        })
    }
    pub fn parse_tree_sitter(&self, content: &str) -> Result<tree_sitter::Tree, ParseError> {
        let mut parser = Parser::new();
        parser.set_language(&self.tree_sitter_language)?;
        parser.parse(content, None).ok_or(ParseError::Miscelaneous)
    }
    pub fn nodes(&self) -> &Vec<Arc<Node>> {
        &self.nodes
    }
    pub fn root_node(&self) -> String {
        self.nodes()
            .iter()
            .find(|node| node.root)
            .unwrap_or_else(|| panic!("No root node found for language: {}", self.name))
            .type_name
            .to_case(Case::Pascal)
    }
    pub fn kind_id(&self, name: &str, named: bool) -> u16 {
        self.tree_sitter_language.id_for_node_kind(name, named)
    }
    pub fn kind_name(&self, id: u16) -> Option<&'static str> {
        self.tree_sitter_language.node_kind_for_id(id)
    }
    pub fn field_id(&self, name: &str) -> Option<NonZeroU16> {
        self.tree_sitter_language.field_id_for_name(name)
    }
    pub fn field_name(&self, id: u16) -> Option<&'static str> {
        self.tree_sitter_language.field_name_for_id(id)
    }
    pub fn name(&self) -> &'static str {
        self.name
    }
    pub fn struct_name(&self) -> &'static str {
        self.struct_name
    }
    pub fn file_struct_name(&self) -> syn::Ident {
        syn::Ident::new(&format!("{}File", self.struct_name()), Span::call_site())
    }
    pub fn package_name(&self) -> String {
        format!("codegen_sdk_{}", self.name())
    }
    pub fn node_for_struct_name(&self, struct_name: &str) -> Option<Arc<Node>> {
        self.nodes
            .iter()
            .find(|node| normalize_type_name(&node.type_name, node.named) == struct_name)
            .cloned()
    }
    pub fn should_parse(&self, file_path: &PathBuf) -> Result<bool, ParseError> {
        Ok(self.file_extensions.contains(
            &file_path
                .extension()
                .ok_or(ParseError::Miscelaneous)?
                .to_str()
                .ok_or(ParseError::Miscelaneous)?,
        ))
    }
}
#[cfg(feature = "go")]
pub mod go;
#[cfg(feature = "java")]
pub mod java;
#[cfg(feature = "typescript")]
pub mod javascript;
#[cfg(feature = "json")]
pub mod json;
#[cfg(feature = "typescript")]
pub mod jsx;
#[cfg(feature = "markdown")]
pub mod markdown;
#[cfg(feature = "python")]
pub mod python;
#[cfg(feature = "ruby")]
pub mod ruby;
#[cfg(feature = "rust")]
pub mod rust;
#[cfg(feature = "toml")]
pub mod toml;
#[cfg(feature = "ts_query")]
pub mod ts_query;
#[cfg(feature = "typescript")]
pub mod tsx;
#[cfg(feature = "typescript")]
pub mod typescript;
#[cfg(feature = "yaml")]
pub mod yaml;
lazy_static! {
    pub static ref LANGUAGES: Vec<&'static Language> = vec![
        #[cfg(feature = "python")]
        &python::Python,
        #[cfg(feature = "typescript")]
        &typescript::Typescript,
        #[cfg(feature = "typescript")]
        &tsx::TSX,
        #[cfg(feature = "typescript")]
        &jsx::JSX,
        #[cfg(feature = "typescript")]
        &javascript::Javascript,
        #[cfg(feature = "rust")]
        &rust::Rust,
        #[cfg(feature = "go")]
        &go::Go,
        #[cfg(feature = "ruby")]
        &ruby::Ruby,
        #[cfg(feature = "yaml")]
        &yaml::Yaml,
        #[cfg(feature = "toml")]
        &toml::TOML,
        #[cfg(feature = "markdown")]
        &markdown::Markdown,
        #[cfg(feature = "json")]
        &json::JSON,
        #[cfg(feature = "java")]
        &java::Java,
        #[cfg(feature = "ts_query")]
        &ts_query::Query,
    ];
}
