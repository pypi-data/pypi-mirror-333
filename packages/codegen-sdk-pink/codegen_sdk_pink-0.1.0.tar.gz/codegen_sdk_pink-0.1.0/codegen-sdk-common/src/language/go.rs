use super::Language;
lazy_static! {
    pub static ref Go: Language = Language::new(
        "go",
        "Go",
        tree_sitter_go::NODE_TYPES,
        &["go"],
        tree_sitter_go::LANGUAGE.into(),
        tree_sitter_go::TAGS_QUERY,
    )
    .unwrap();
}
