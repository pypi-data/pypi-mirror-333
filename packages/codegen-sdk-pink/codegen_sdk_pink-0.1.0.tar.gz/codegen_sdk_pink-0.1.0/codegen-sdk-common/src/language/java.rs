use super::Language;
lazy_static! {
    pub static ref Java: Language = Language::new(
        "java",
        "Java",
        tree_sitter_java::NODE_TYPES,
        &["java"],
        tree_sitter_java::LANGUAGE.into(),
        tree_sitter_java::TAGS_QUERY,
    )
    .unwrap();
}
