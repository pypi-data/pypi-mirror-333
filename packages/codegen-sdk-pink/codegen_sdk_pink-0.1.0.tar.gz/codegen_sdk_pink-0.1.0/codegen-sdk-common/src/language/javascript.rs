use super::Language;

lazy_static! {
    pub static ref Javascript: Language = Language::new(
        "javascript",
        "Javascript",
        tree_sitter_javascript::NODE_TYPES,
        &["js"],
        tree_sitter_javascript::LANGUAGE.into(),
        tree_sitter_javascript::TAGS_QUERY,
    )
    .unwrap();
}
