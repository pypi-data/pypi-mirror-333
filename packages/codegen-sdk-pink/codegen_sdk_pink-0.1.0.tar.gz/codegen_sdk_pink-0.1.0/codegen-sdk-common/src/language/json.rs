use super::Language;
lazy_static! {
    pub static ref JSON: Language = Language::new(
        "json",
        "JSON",
        tree_sitter_json::NODE_TYPES,
        &["json"],
        tree_sitter_json::LANGUAGE.into(),
        "",
    )
    .unwrap();
}
