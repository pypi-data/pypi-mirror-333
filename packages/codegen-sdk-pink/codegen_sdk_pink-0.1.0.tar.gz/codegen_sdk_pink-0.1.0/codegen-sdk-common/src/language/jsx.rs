use super::Language;
lazy_static! {
    pub static ref JSX: Language = Language::new(
        "jsx",
        "JSX",
        tree_sitter_typescript::TSX_NODE_TYPES,
        &["jsx"],
        tree_sitter_typescript::LANGUAGE_TSX.into(),
        tree_sitter_typescript::TAGS_QUERY,
    )
    .unwrap();
}
