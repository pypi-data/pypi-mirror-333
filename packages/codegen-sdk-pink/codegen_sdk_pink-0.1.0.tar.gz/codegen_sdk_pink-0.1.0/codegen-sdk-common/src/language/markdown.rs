use super::Language;
lazy_static! {
    pub static ref Markdown: Language = Language::new(
        "markdown",
        "Markdown",
        tree_sitter_md::NODE_TYPES_BLOCK,
        &["md"],
        tree_sitter_md::LANGUAGE.into(),
        "",
    )
    .unwrap();
}
