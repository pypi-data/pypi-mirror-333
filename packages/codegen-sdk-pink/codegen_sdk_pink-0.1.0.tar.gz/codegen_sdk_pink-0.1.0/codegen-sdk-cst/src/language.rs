use std::{hash::Hash, path::PathBuf, sync::Arc};

use bytes::Bytes;
use codegen_sdk_common::{
    ParseError, Tree,
    language::Language,
    traits::{CSTNode, FromNode},
    tree::TreeNode,
};

pub trait CSTLanguage {
    type Types<'db>: TreeNode;
    type Program<'db1>: CSTNode<'db1> + FromNode<'db1, Self::Types<'db1>> + Send;
    fn language() -> &'static Language;
    fn parse<'db>(
        db: &'db dyn salsa::Database,
        content: String,
    ) -> Option<(
        &'db Self::Program<'db>,
        &'db Tree<Self::Types<'db>>,
        indextree::NodeId,
    )>;
    fn parse_file_from_cache<'db>(
        db: &'db dyn salsa::Database,
        file_path: &PathBuf,
        #[cfg(feature = "serialization")] cache: &'db codegen_sdk_common::serialize::Cache,
    ) -> Result<
        Option<(
            &'db Self::Program<'db>,
            &'db Tree<Self::Types<'db>>,
            indextree::NodeId,
        )>,
        ParseError,
    > {
        #[cfg(feature = "serialization")]
        {
            let serialized_path = cache.get_path(file_path);
            if serialized_path.exists() {
                let parsed = cache.read_entry::<Self::Program<'db>>(&serialized_path)?;
                return Ok(Some(parsed));
            }
        }
        Ok(None)
    }
    fn parse_file<'db>(
        db: &'db dyn salsa::Database,
        file_path: &PathBuf,
        #[cfg(feature = "serialization")] cache: &'db codegen_sdk_common::serialize::Cache,
    ) -> Result<
        Option<(
            &'db Self::Program<'db>,
            &'db Tree<Self::Types<'db>>,
            indextree::NodeId,
        )>,
        ParseError,
    > {
        if let Some(parsed) = Self::parse_file_from_cache(
            db,
            file_path,
            #[cfg(feature = "serialization")]
            cache,
        )? {
            return Ok(Some(parsed));
        }
        let content = std::fs::read_to_string(file_path)?;
        if let Some(parsed) = Self::parse(db, content) {
            return Ok(Some(parsed));
        }
        Err(ParseError::SyntaxError)
    }

    fn should_parse(file_path: &PathBuf) -> Result<bool, ParseError> {
        Self::language().should_parse(file_path)
    }
}
