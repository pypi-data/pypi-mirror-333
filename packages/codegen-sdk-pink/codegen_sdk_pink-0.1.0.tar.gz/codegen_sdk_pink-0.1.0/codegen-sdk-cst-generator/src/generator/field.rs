// Representation of a field of a node. Can be optional, required, or, multiple
use std::collections::HashSet;

#[double]
use codegen_sdk_common::language::Language;
use codegen_sdk_common::{
    naming::{field_name_getter, normalize_field_name, normalize_type_name},
    parser::{FieldDefinition, TypeDefinition},
};
use mockall_double::double;
use proc_macro2::TokenStream;
use quote::{format_ident, quote};
use syn::parse_quote;

use crate::Config;

#[derive(Debug)]
pub struct Field<'a> {
    raw: FieldDefinition,
    name: String,
    node_name: String,
    language: &'a Language,
    config: Config,
}

impl<'a> Field<'a> {
    pub fn new(
        node_name: &str,
        name: &str,
        raw: FieldDefinition,
        language: &'a Language,
        config: Config,
    ) -> Self {
        Self {
            node_name: node_name.to_string(),
            name: name.to_string(),
            raw,
            language,
            config,
        }
    }
    fn field_id(&self) -> u16 {
        self.language.field_id(&self.name).unwrap().into()
    }
    pub fn name(&self) -> String {
        normalize_field_name(&self.name)
    }
    pub fn normalized_name(&self) -> String {
        normalize_type_name(&self.name, true)
    }
    pub fn types(&self) -> Vec<&TypeDefinition> {
        self.raw.types.iter().collect()
    }
    pub fn type_name(&self) -> String {
        let types = self.types();
        if types.len() == 1 {
            normalize_type_name(&types[0].type_name, types[0].named)
        } else {
            format!("{}{}", self.node_name, self.normalized_name())
        }
    }
    pub fn type_name_ref(&self, subenums: &HashSet<&String>) -> TokenStream {
        let types = self.types();
        // We return a direct reference if the there is only one type and it is not a subenum
        if types.len() == 1 {
            let ident = format_ident!("{}", types[0].normalize());
            if subenums.contains(&types[0].type_name) {
                let ident = format_ident!("{}Ref", ident);
                quote! {#ident}
            } else {
                quote! {
                    &#ident
                }
            }
        } else {
            let ident = format_ident!("{}{}Ref", self.node_name, self.normalized_name());
            quote! {
                #ident
            }
        }
    }

    pub fn get_constructor_field(&self) -> TokenStream {
        let field_name_ident = format_ident!("{}", self.name());
        let original_name = &self.name;
        let converted_type_name = format_ident!("{}", self.type_name());
        if self.raw.multiple {
            quote! {
                let #field_name_ident= get_multiple_children_by_field_name::<NodeTypes<'db>, #converted_type_name<'db>>(context, &node, #original_name)?;
                for child in #field_name_ident.iter().cloned() {
                    ids.push(child);
                }
            }
        } else if !self.raw.required {
            quote! {
                let #field_name_ident = get_optional_child_by_field_name::<NodeTypes<'db>, #converted_type_name<'db>>(context, &node, #original_name)?;
                if let Some(child) = #field_name_ident.clone() {
                    ids.push(child);
                }
            }
        } else {
            quote! {
                let #field_name_ident = get_child_by_field_name::<NodeTypes<'db>, #converted_type_name<'db>>(context, &node, #original_name)?;
                ids.push(#field_name_ident.clone());
            }
        }
    }
    pub fn get_convert_child(&self, convert_children: bool) -> TokenStream {
        let field_name_ident = format_ident!("{}", self.name());
        if convert_children {
            if self.raw.multiple {
                quote! {
                    context.get(child).unwrap().as_ref().try_into().unwrap()
                }
            } else if !self.raw.required {
                quote! {
                    context.get(child).unwrap().as_ref().try_into().unwrap()
                }
            } else {
                quote! {
                    context.get(&self.#field_name_ident).unwrap().as_ref().try_into().unwrap()
                }
            }
        } else if self.raw.multiple || !self.raw.required {
            quote! {
                child
            }
        } else {
            quote! {
                self.#field_name_ident.as_ref()
            }
        }
    }
    pub fn get_children_field(&self, convert_children: bool) -> TokenStream {
        let field_name_ident = format_ident!("{}", self.name());
        let convert_child = self.get_convert_child(convert_children);

        if self.raw.multiple {
            quote! {
                children.extend(self.#field_name_ident.iter().map(|child| #convert_child));
            }
        } else if self.raw.required {
            quote! {
                children.push(#convert_child);
            }
        } else {
            quote! {
                if let Some(child) = self.#field_name_ident.as_ref() {
                    children.push(#convert_child);
                }
            }
        }
    }
    pub fn get_children_by_field_name_field(&self, convert_children: bool) -> TokenStream {
        let field_name = &self.name;
        let field_name_ident = format_ident!("{}", self.name());
        let convert_child = self.get_convert_child(convert_children);

        if self.raw.multiple {
            quote! {
                #field_name => self.#field_name_ident.iter().map(|child| #convert_child).collect()
            }
        } else if self.raw.required {
            quote! {
                #field_name => vec![#convert_child]
            }
        } else {
            quote! {
                #field_name => self.#field_name_ident.as_ref().iter().map(|child| #convert_child).collect()
            }
        }
    }
    pub fn get_children_by_field_id_field(&self, convert_children: bool) -> TokenStream {
        let field_id = self.field_id();
        let field_name_ident = format_ident!("{}", self.name());
        let convert_child = self.get_convert_child(convert_children);

        if self.raw.multiple {
            quote! {
                #field_id => self.#field_name_ident.iter().map(|child| #convert_child).collect()
            }
        } else if self.raw.required {
            quote! {
                #field_id => vec![#convert_child]
            }
        } else {
            quote! {
                #field_id => self.#field_name_ident.as_ref().iter().map(|child| #convert_child).collect()
            }
        }
    }
    pub fn get_struct_field(&self) -> TokenStream {
        let field_name_ident = format_ident!("{}", self.name());
        let bounds = if self.config.serialize {
            quote! {
                #[rkyv(omit_bounds)]
            }
        } else {
            quote! {}
        };
        if self.raw.multiple {
            quote! {
                #bounds
                pub #field_name_ident: Vec<indextree::NodeId>
            }
        } else if !self.raw.required {
            quote! {
                #bounds
                pub #field_name_ident: Option<indextree::NodeId>
            }
        } else {
            quote! {
                #bounds
                pub #field_name_ident: indextree::NodeId
            }
        }
    }
    pub fn get_field_getter(&self, subenums: &HashSet<&String>) -> syn::ImplItemFn {
        let function_name = format_ident!("{}", field_name_getter(&self.name()));
        let field_name_ident = format_ident!("{}", self.name());
        let converted_type_name = self.type_name_ref(subenums);
        if self.raw.multiple {
            parse_quote! {
                pub fn #function_name(&self, tree: &'db Tree<NodeTypes<'db>>) -> Vec<#converted_type_name<'db>> {
                    self.#field_name_ident.iter().map(|id| tree.get(id).unwrap().as_ref().try_into().unwrap()).collect()
                }
            }
        } else if !self.raw.required {
            parse_quote! {
                pub fn #function_name(&self, tree: &'db Tree<NodeTypes<'db>>) -> Option<#converted_type_name<'db>> {
                    if let Some(id) = self.#field_name_ident {
                        Some(tree.get(&id).unwrap().as_ref().try_into().unwrap())
                    } else {
                        None
                    }
                }
            }
        } else {
            parse_quote! {
                pub fn #function_name(&self, tree: &'db Tree<NodeTypes<'db>>) -> #converted_type_name<'db> {
                    tree.get(&self.#field_name_ident).unwrap().as_ref().try_into().unwrap()
                }
            }
        }
    }

    pub fn is_optional(&self) -> bool {
        !self.raw.required
    }
    pub fn is_multiple(&self) -> bool {
        self.raw.multiple
    }
}

#[cfg(test)]
mod tests {
    use codegen_sdk_common::parser::TypeDefinition;

    use super::*;
    use crate::test_util::{get_language_no_nodes, snapshot_tokens};
    fn create_test_field_definition(name: &str, multiple: bool, required: bool) -> FieldDefinition {
        FieldDefinition {
            types: vec![TypeDefinition {
                type_name: name.to_string(),
                named: true,
            }],
            multiple,
            required,
        }
    }
    fn create_test_field_definition_variants(
        name: &Vec<String>,
        multiple: bool,
        required: bool,
    ) -> FieldDefinition {
        FieldDefinition {
            types: name
                .iter()
                .map(|n| TypeDefinition {
                    type_name: n.to_string(),
                    named: true,
                })
                .collect(),
            multiple,
            required,
        }
    }

    #[test]
    fn test_field_normalized_name() {
        let field_definition = create_test_field_definition("test_type", false, true);
        let language = get_language_no_nodes();
        let field = Field::new(
            "node",
            "field",
            field_definition,
            &language,
            Config::default(),
        );
        assert_eq!(field.normalized_name(), "Field");
    }

    #[test]
    fn test_field_types() {
        let field_definition = create_test_field_definition_variants(
            &vec!["type_a".to_string(), "type_b".to_string()],
            false,
            true,
        );
        let language = get_language_no_nodes();
        let field = Field::new(
            "test_node",
            "test_field",
            field_definition.clone(),
            &language,
            Config::default(),
        );
        assert_eq!(
            field.types(),
            field_definition.types.iter().collect::<Vec<_>>()
        );
    }

    #[test]
    fn test_field_type_name() {
        let field_definition = create_test_field_definition_variants(
            &vec!["test_type".to_string(), "test_type".to_string()],
            false,
            true,
        );
        let language = get_language_no_nodes();
        let field = Field::new(
            "Node",
            "field",
            field_definition,
            &language,
            Config::default(),
        );
        assert_eq!(field.type_name(), "NodeField");
    }

    #[test]
    fn test_get_struct_field() {
        let field_definition = create_test_field_definition("test_type", false, true);
        let language = get_language_no_nodes();
        let field = Field::new(
            "test_node",
            "test_field",
            field_definition,
            &language,
            Config::default(),
        );
        insta::assert_debug_snapshot!(snapshot_tokens(&field.get_struct_field()));

        // Test optional field
        let optional_definition = create_test_field_definition("test_type", false, false);
        let optional_field = Field::new(
            "test_node",
            "test_field",
            optional_definition,
            &language,
            Config::default(),
        );
        insta::assert_debug_snapshot!(snapshot_tokens(&optional_field.get_struct_field()));

        // Test multiple field
        let multiple_definition = create_test_field_definition("test_type", true, true);
        let multiple_field = Field::new(
            "test_node",
            "test_field",
            multiple_definition,
            &language,
            Config::default(),
        );
        insta::assert_debug_snapshot!(snapshot_tokens(&multiple_field.get_struct_field()));
    }

    #[test]
    fn test_get_constructor_field() {
        let field_definition = create_test_field_definition("test_type", false, true);
        let language = get_language_no_nodes();
        let field = Field::new(
            "test_node",
            "test_field",
            field_definition,
            &language,
            Config::default(),
        );
        insta::assert_debug_snapshot!(snapshot_tokens(&field.get_constructor_field()));

        // Test optional field
        let optional_definition = create_test_field_definition("test_type", false, false);
        let optional_field = Field::new(
            "test_node",
            "test_field",
            optional_definition,
            &language,
            Config::default(),
        );
        insta::assert_debug_snapshot!(snapshot_tokens(&optional_field.get_constructor_field()));

        // Test multiple field
        let multiple_definition = create_test_field_definition("test_type", true, true);
        let multiple_field = Field::new(
            "test_node",
            "test_field",
            multiple_definition,
            &language,
            Config::default(),
        );
        insta::assert_debug_snapshot!(snapshot_tokens(&multiple_field.get_constructor_field()));
    }

    #[test]
    fn test_get_children_field() {
        let field_definition = create_test_field_definition("test_type", false, true);
        let language = get_language_no_nodes();
        let field = Field::new(
            "test_node",
            "test_field",
            field_definition,
            &language,
            Config::default(),
        );

        assert_eq!(
            field.get_children_field(true).to_string(),
            quote!(children.push(context.get(&self.test_field).unwrap().as_ref().try_into().unwrap());).to_string()
        );

        // Test optional field
        let optional_definition = create_test_field_definition("test_type", false, false);
        let optional_field = Field::new(
            "test_node",
            "test_field",
            optional_definition,
            &language,
            Config::default(),
        );

        assert_eq!(
            optional_field.get_children_field(true).to_string(),
            quote!(if let Some(child) = self.test_field.as_ref() {
                children.push(context.get(child).unwrap().as_ref().try_into().unwrap());
            })
            .to_string()
        );

        // Test multiple field
        let multiple_definition = create_test_field_definition("test_type", true, true);
        let multiple_field = Field::new(
            "test_node",
            "test_field",
            multiple_definition,
            &language,
            Config::default(),
        );

        assert_eq!(
            multiple_field.get_children_field(true).to_string(),
            quote!(children.extend(self.test_field.iter().map(|child| context.get(child).unwrap().as_ref().try_into().unwrap()));).to_string()
        );
    }

    #[test]
    fn test_get_children_by_field_name_field() {
        let field_definition = create_test_field_definition("test_type", false, true);
        let language = get_language_no_nodes();
        let field = Field::new(
            "test_node",
            "test_field",
            field_definition,
            &language,
            Config::default(),
        );

        assert_eq!(
            field.get_children_by_field_name_field(true).to_string(),
            quote!("test_field" => vec![context.get(&self.test_field).unwrap().as_ref().try_into().unwrap()]).to_string()
        );

        // Test optional field
        let optional_definition = create_test_field_definition("test_type", false, false);
        let optional_field = Field::new(
            "test_node",
            "test_field",
            optional_definition,
            &language,
            Config::default(),
        );

        assert_eq!(
            optional_field.get_children_by_field_name_field(true).to_string(),
            quote!("test_field" => self.test_field.as_ref().iter().map(|child| context.get(child).unwrap().as_ref().try_into().unwrap()).collect()).to_string()
        );

        // Test multiple field
        let multiple_definition = create_test_field_definition("test_type", true, true);
        let multiple_field = Field::new(
            "test_node",
            "test_field",
            multiple_definition,
            &language,
            Config::default(),
        );

        assert_eq!(
            multiple_field.get_children_by_field_name_field(true).to_string(),
            quote!("test_field" => self.test_field.iter().map(|child| context.get(child).unwrap().as_ref().try_into().unwrap()).collect()).to_string()
        );
    }
}
