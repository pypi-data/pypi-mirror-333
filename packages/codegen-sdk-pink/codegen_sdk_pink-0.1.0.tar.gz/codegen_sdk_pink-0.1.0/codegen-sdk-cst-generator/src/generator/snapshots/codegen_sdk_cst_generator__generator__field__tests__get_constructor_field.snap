---
source: codegen-sdk-cst-generator/src/generator/field.rs
expression: snapshot_tokens(&field.get_constructor_field())
---
let test_field = get_child_by_field_name :: < NodeTypes < 'db > , TestType < 'db >> (context , & node , "test_field") ? ; ids . push (test_field . clone ()) ;
